topdown/input.go

```diff
@@ -43,21 +43,23 @@ func mergeTermWithValues(exist *ast.Term, pairs [][2]*ast.Term) (*ast.Term, erro
 				result = exist.Copy()
 				init = true
 			}
-
 			if result == nil {
 				result = ast.NewTerm(makeTree(target[1:], pair[1]))
 			} else {
 				node := result
 				done := false
 				for i := 1; i < len(target)-1 && !done; i++ {
-					if child := node.Get(target[i]); child == nil {
-						obj, ok := node.Value.(ast.Object)
-						if !ok {
-							return nil, errConflictingDoc
-						}
+					obj, ok := node.Value.(ast.Object)
+					if !ok {
+						return nil, errConflictingDoc // possible?
+					}
+					if child := obj.Get(target[i]); child == nil {
+						obj.Insert(target[i], ast.NewTerm(makeTree(target[i+1:], pair[1])))
+						done = true
+					} else if _, ok := child.Value.(ast.Object); !ok {
```

Comment:
nit: we could combine this into one branch by adding a helper function like `isObject(x *ast.Term) bool` that returns true if x.Value.(ast.Object) is true, otherwise false (including if x is nil).

Reply:
Yeah, it hadn't bothered me as much when writing; looking at those lines this morning, combining them seems obvious 😄 Done.